docs/proposals/multiple-sources-for-applications.md

```diff
@@ -0,0 +1,234 @@
+```
+
+Creating applications from multiple sources would allow users to configure multiple services stored at various sources within the same application.
+
+## Motivation
+
+The main motivation behind this enhancement proposal is to allow users to create an application using services that are stored in various sources.
+
+### Goals
+
+The goals of the enhancement are:
+
+#### **Supporting multiple sources for creating an application**
+
+Users should be able to specify multiple sources to add services to the application. Argo CD should compile all the sources and reconcile each source individually for creating the application.
+
+#### **Allow specifying external value files for Helm repositories**
+
+Users should be able to specify different sources for Helm charts and values files. The Helm charts specified by the user could be available in Git or Helm repository and the value files are stored in Git. Argo CD should track changes in both the Helm charts and the value files repository and reconcile the application.
+
+#### Changes to UI
+
+The UI should allow users to add multiple sources while creating the application. For each source, UI should allow users to add multiple external values files Helm projects. We would need a separate proposal for changes to UI.
+
+#### Changes to cli
+
+The cli would need to support adding a list of resources instead of just one while creating the application. Also, just like `--values` field for adding values files, we would need an option to allow external value files to the application. We would need a separate proposal for changes to cli.
+
+### Non-goals
+* 
+
+## Proposal
+
+### Add new `sources` field in Application Spec
+
+The proposal is to add a new field `sources` which would allow users to input list of `ApplicationSource`s. We would mark field `source` as deprecated and would ignore the details under `source` with details under `sources` field. 
+
+Below example shows how the yaml would look like for `source` and `sources` field. We would ignore the `source` field and apply the resources mentioned under `sources` field.
+
+```yaml
+spec:
+  source:
+    repoURL: https://github.com/elastic/helm-charts/tree/main/elasticsearch
+    targetRevision: 6.8
+    helm:
+      valueFiles:
+        - values.yaml
+    chart: elasticsearch
+  sources:                                          # new field
+    - repoURL: https://github.com/helm/charts/tree/master/incubator/elasticsearch
+      targetRevision: master
+      helm:
+        valueFiles:
+          - values.yaml
+      chart: elasticsearch
+```
+
+### Add `externalValues` field in helm section of Application Spec
```

Comment:
Unless I'm missing something, the proposed design allows using additional sources only with Helm. However I think we have two common use cases here that can be useful for any config management tool:

* Provide additional set of file that are needed to generate manifests ( like values.yaml for helm template )
* Create an application that is combined with several helm charts/kustomizes. Right now users are forced to create several apps or use some config management tool to combine two sources. 


First use case is easier. We just let use define multiple sources. In this case Argo CD would generate manifests from all sources and merge them in one list:

```yaml
spec:
  # application that consists of MongoDB and ElasticSearch resources
  sources:
    - repoURL: https://github.com/helm/charts
      targetRevision: master
      path: incubator/mongodb
    - repoURL: https://github.com/helm/charts
      targetRevision: master
      path: incubator/elasticsearch
``` 

To enable the second use case I propose to make `path/chart` field optional: if it is missing then manifests are not generated for the source. In order to make repo files accessible for other sources user would have to provide a `ref` that might be referenced in other sources.

```yaml
spec:
  sources:
    - repoURL: https://github.com/my-org/my-repo # path is missing so no manifests are generated
      targetRevision: master
      ref: myRepo                                 # repo is available via symlink "my-repo"
    - repoURL: https://github.com/helm/charts
      targetRevision: master
      path: incubator/elasticsearch               # path "incubator/elasticsearch" is used to generate manifests
      helm:
        valueFiles:
          - $myRepo/values.yaml                   # values.yaml is referenced in firstSource 
```

This way we can cover Helm use-case without introducing Helm specific settings and other use cases as well. WDYT?

Reply:
Hi @alexmt, the idea above sounds good. I just had a small question.

For the second use case, I was thinking how adding multiple files would work. For example, (expanding on your example above)

```
spec:
  sources:
    - repoURL: https://github.com/my-org/my-repo # path is missing so no manifests are generated
      targetRevision: master
      ref: myRepo                                 # repo is available via symlink "my-repo"
    - repoURL: https://github.com/helm/charts
      targetRevision: master
      path: incubator/elasticsearch               # path "incubator/elasticsearch" is used to generate manifests
      ref: repoES
      helm:
        valueFiles:
          - $myRepo/values.yaml                   # values.yaml is located in source with reference name $myRepo
          - gen_values.yaml                       # values file present in the path directory of current source
          - $repoES/esValues.yaml                 # values file present in the same path directory of current source
```
Would we support adding values files from current source using it's ref `$repoES` (e.g. `$repoES/esValues.yaml`) and also without it's ref (e.g. `gen_values.yaml`)?

Also, would the referenced source (for e.g. `$myRepo`) need to be declared before using them in other sources?

Reply:
I'm thinking the Argo CD would just resolve `$<refName>` to the absolute path where the repo is cloned. To reference values file from current source user would just use a relative path (same as now). The source needs reference only if some other source needs to use file from it.

Reply:
It would be REALLY nice if value file "pointers" also could have a branch (like the application itself has) - so for development f.ex. - you'd point an application at a 'testing' branc - and the values file then also to an equivalent testing branch in its repo.
Thats how we typicly test applications before we merge them - by adding them - pointing to the devel branch for testing via argocd - but today - our values files HAVE to be in master branch - as we use the path hack from above (where the repo is in /tmp/.../ - because its a repo we already have added).

Reply:
@KlavsKlavsen this should be covered by the `targetRevision` field. The only caveat is that it won't be possible to reference two revisions of the same branch. This limitation will be removed after switching to Config Management Plugins though.